pidTuner.cpp

```diff
@@ -0,0 +1,109 @@
+#include "pidTuner.hpp"
+#include <iostream>
+#include <math.h>
+
+
+pidTuner::pidTuner(float ip, float ii, float id, float Sp, float Si, float Sd){
+    _initial_pid = pid_set(ip,ii,id);
+    _current_pid =  _initial_pid;
+    _hill_vector = pid_set(1,1,1);
+
+    _cycles = 0;
+    _test_num = 0;
+
+    _pScale = Sp;
+    _iScale = Si;
+    _dScale = Sd;
+
+    _overScale = 1;
+
+    _threshold = .5;
+}
+
+pidTuner::pid_set::pid_set(float ip, float ii, float id){
+    p=ip;
+    i=ii;
+    d=id;
+    score=0;
+}
+
+void pidTuner::start_cycle(){
+    //determine values to test
+    if(_test_num==_test_sets.size()){
+         _test_sets.clear();
+
+        for(int i=0;i<4;i++){
+            pid_set _pid = pid_set(_current_pid.p,_current_pid.i,_current_pid.d);
+
+            switch(i){
+                case 0:
+                    break;
+                case 1:
+                    _pid.p += _pScale;
+                    break;
+                case 2:
+                    _pid.i += _iScale;
+                    break;
+                case 3:
+                    _pid.d += _dScale;
+                    break;
+                }
+                _test_sets.push_back(_pid);
+        }
+
+        _test_num = 0;
+    }
+    _current_pid = _test_sets[_test_num];
+}
+
+void pidTuner::run(float err){
+    _current_pid.score += fabs(err);
+}
+
+//returns if pid needs more tuning, if it does, sets up next test points
+bool pidTuner::end_cycle(){
+    //score old value
+    _test_sets[_test_num] = _current_pid;
+
+    _test_num += 1;
+
+    if(_test_num==_test_sets.size()){
+        //finished testing set, determine best pid, then check if we need more tests
+        _cycles+=1;
+
+
+        pid_set _best_pid = _test_sets[0];
+        for(int i=0; i<_test_sets.size();i++){
+            if(_test_sets[i].score<_best_pid.score){
+                _best_pid = _test_sets[i];
+            }
+        }
+
+        if(_best_pid == _test_sets[0]){
+            _overScale = 2;
```

Comment:
basically, if the pid tuner has overshot in every direction then it divides its stepsize by 2 and starts going back the other way. As for why its a class variable, because I had originally planned to mess around to see what size worked best and then completely forgot about it. If halfing the step size seems good to you guys I can just make it not a class variable